rustls/src/builder.rs

```diff
@@ -62,81 +61,70 @@ use crate::versions;
+///    [`ConfigBuilder<S, WantsKxGroups>::with_safe_default_kx_groups()`].
 /// 3. Now you must make
 ///    a decision on which protocol versions to support, typically by calling
-///    [`ConfigWantsVersions::with_safe_default_protocol_versions()`].
-/// 4. You now need to indicate whether to make a [`ServerConfig`] or [`ClientConfig`],
-///    by calling [`ConfigWantsPeerType::for_server()`]
-///    or [`ConfigWantsPeerType::for_client()`] respectively.
-/// 5. Now see [`ConfigWantsServerVerifier`] or [`ConfigWantsClientVerifier`] for further steps.
+///    [`ConfigBuilder<S, WantsVersions>::with_safe_default_protocol_versions()`].
+/// 5. Now see [`ConfigBuilder<ClientConfig, WantsVerifier>`] or
+///    [`ConfigBuilder<ServerConfig, WantsVerifier>`] for further steps.
 ///
 /// [`ServerConfig`]: crate::ServerConfig
 /// [`ClientConfig`]: crate::ClientConfig
-pub fn config_builder() -> ConfigWantsCipherSuites {
-    ConfigWantsCipherSuites {}
+/// [`ClientConfig::builder()`]: crate::ClientConfig::builder()
+/// [`ServerConfig::builder()`]: crate::ServerConfig::builder()
+/// [`ConfigWantsServerVerifier`]: crate::ConfigWantsServerVerifier
+/// [`ConfigWantsClientVerifier`]: crate::ConfigWantsClientVerifier
+#[derive(Clone)]
+pub struct ConfigBuilder<Side: ConfigSide, State> {
+    pub(crate) state: State,
+    pub(crate) side: PhantomData<Side>,
 }
 
-fn config_builder_with_safe_defaults() -> ConfigWantsPeerType {
-    config_builder()
-        .with_safe_default_cipher_suites()
-        .with_safe_default_kx_groups()
-        .with_safe_default_protocol_versions()
-}
+/// Config builder state where the caller must supply cipher suites.
+#[derive(Clone)]
+pub struct WantsCipherSuites;
 
-/// Start building a [`ClientConfig`] and accept defaults for underlying
-/// cryptography.
-///
-/// These are safe defaults, useful for 99% of applications.
-///
-/// [`ClientConfig`]: crate::ClientConfig
-pub fn client_config_builder_with_safe_defaults() -> ConfigWantsServerVerifier {
-    // this function exists to express that for_client is infallible when
-    // using defaults.
-    config_builder_with_safe_defaults()
-        .for_client()
-        .unwrap()
-}
-
-/// Start building a [`ServerConfig`] and accept defaults for underlying
-/// cryptography.
-///
-/// These are safe defaults, useful for 99% of applications.
-///
-/// [`ServerConfig`]: crate::ServerConfig
-pub fn server_config_builder_with_safe_defaults() -> ConfigWantsClientVerifier {
-    // this function exists to express that for_server is infallible when
-    // using defaults.
-    config_builder_with_safe_defaults()
-        .for_server()
-        .unwrap()
-}
-
-/// A config builder where we want to know the cipher suites.
-pub struct ConfigWantsCipherSuites;
+impl<S: ConfigSide> ConfigBuilder<S, WantsCipherSuites> {
+    /// Start side-specific config with defaults for underlying cryptography.
+    ///
+    /// These are safe defaults, useful for 99% of applications.
```

Comment:
this is not directly related to the change made here, since the doc comment was already present in the previous code, but it seems like it might be nice to document what the defaults _are_?